File: common/src/main/java/com/ssblur/scriptor/word/subject/HitscanSubject.java
package com.ssblur.scriptor.word.subject;

import com.ssblur.scriptor.api.word.Subject;
import com.ssblur.scriptor.events.network.TraceNetwork;
import com.ssblur.scriptor.helpers.targetable.EntityTargetable;
import com.ssblur.scriptor.helpers.targetable.LecternTargetable;
import com.ssblur.scriptor.helpers.targetable.Targetable;
import com.ssblur.scriptor.word.Spell;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.Vec3;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class HitscanSubject extends Subject {

  @Override
  public CompletableFuture<List<Targetable>> getTargets(Targetable caster, Spell spell) {
    var result = new CompletableFuture<List<Targetable>>();
    if(caster instanceof EntityTargetable entityTargetable && entityTargetable.getTargetEntity() instanceof Player player) {
      TraceNetwork.requestExtendedTraceData(player, target -> result.complete(List.of(target)));
    } else if(caster instanceof LecternTargetable) {
      var pos = caster.getTargetBlockPos().relative(caster.getFacing());
      var normal = caster.getFacing().getNormal();
      int x = normal.getX() * 19;
      int z = normal.getZ() * 19;
      var entities = caster.getLevel().getEntitiesOfClass(
        LivingEntity.class,
        AABB.ofSize(
          new Vec3(
            pos.getX(),
            pos.getY(),
            pos.getZ()
          ),
          1 + x,
          1,
          1 + z
        )
      );

      if(entities.isEmpty()) {
        int i = 0;
        x = caster.getFacing().getStepX();
        z = caster.getFacing().getStepZ();
        while(i < 19) {
          if(
            !caster.getLevel().getBlockState(
              pos.offset(x * i, 0, z * i)
            ).isAir()
          ) {
            result.complete(List.of(new Targetable(caster.getLevel(), pos.offset(x * i, 0, z * i))));
            return result;
          }
          i++;
        }

      } else
        result.complete(entities.stream().map(living -> (Targetable) new EntityTargetable(living)).toList());
    } else {
      result.complete(List.of(caster.simpleCopy()));
    }
    return result;
  }

  @Override
  public Cost cost() { return new Cost(8, COSTTYPE.ADDITIVE); }
}

File: common/src/main/kotlin/com/ssblur/scriptor/word/subject/HitscanSubject.kt
package com.ssblur.scriptor.word.subject

import com.ssblur.scriptor.api.word.Subject
import com.ssblur.scriptor.color.CustomColors.getColor
import com.ssblur.scriptor.helpers.targetable.EntityTargetable
import com.ssblur.scriptor.helpers.targetable.LecternTargetable
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.network.client.ParticleNetwork
import com.ssblur.scriptor.network.server.TraceNetwork
import com.ssblur.scriptor.word.Spell
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import java.util.concurrent.CompletableFuture

class HitscanSubject : Subject() {
    override fun getTargets(caster: Targetable, spell: Spell): CompletableFuture<List<Targetable>> {
        val result = CompletableFuture<List<Targetable>>()
        if (caster is EntityTargetable && caster.targetEntity is Player) {
            val player = caster.targetEntity as Player
            TraceNetwork.requestExtendedTraceData(player) { target: Targetable ->
                val color = getColor(spell.deduplicatedDescriptorsForSubjects())
                ParticleNetwork.magicTrail(target.level, color, player.eyePosition, target.targetPos)
                result.complete(listOf(target))
            }
        } else if (caster is LecternTargetable) {
            val pos = caster.targetBlockPos.relative(caster.facing)
            val normal = caster.facing.normal
            var x = normal.x * 19
            var z = normal.z * 19
            val entities = caster.level.getEntitiesOfClass(
                LivingEntity::class.java,
                AABB.ofSize(
                    Vec3(
                        pos.x.toDouble(),
                        pos.y.toDouble(),
                        pos.z.toDouble()
                    ),
                    (1 + x).toDouble(),
                    1.0,
                    (1 + z).toDouble()
                )
            )

            if (entities.isEmpty()) {
                var i = 0
                x = caster.facing.stepX
                z = caster.facing.stepZ
                while (i < 19) {
                    if (!caster.level.getBlockState(pos.offset(x * i, 0, z * i)).isAir) {
                        result.complete(listOf(Targetable(caster.level, pos.offset(x * i, 0, z * i))))
                        return result
                    }
                    i++
                }
            } else result.complete(entities.stream().map { EntityTargetable(it) }.toList())
        } else {
            result.complete(listOf(caster.simpleCopy()))
        }
        return result
    }

    override fun cost(): Cost {
        return Cost(8.0, COSTTYPE.ADDITIVE)
    }
}
